src/diversity/similarity.py

```diff
@@ -195,407 +52,189 @@ def calculate_weighted_similarities(self, relative_abundances, out=None):
+        chunk_size:
+            Determines how many rows of the similarity matrix each will be processes at a time.
+            In general, choosing a larger chunk_size will make the calculation faster,
+            but will also require more memory.
         """
         LOGGER.debug(
-            "__make_shared(relative_abundances=%s, out=%s)", relative_abundances, out
-        )
-        if not isinstance(relative_abundances, SharedArrayView):
-            relative_abundances_ = self.__shared_array_manager.from_array(
-                relative_abundances
-            )
-        else:
-            relative_abundances_ = relative_abundances
-        if out is None or not isinstance(out, SharedArrayView):
-            out_ = self.__shared_array_manager.empty(
-                shape=relative_abundances_.spec.shape, data_type=dtype("f8")
-            )
-        else:
-            out_ = out
-        return (relative_abundances_, out_)
-
-    def calculate_weighted_similarities(self, relative_abundances, out=None):
-        """Same as diversity.metacommunity.ISimilarity.calculate_weighted_similarities, except using shared arrays."""
-        LOGGER.debug(
-            "calculate_weighted_similarities(relative_abundances=%s, out=%s)",
-            relative_abundances,
-            out,
-        )
-        relative_abundances_, out_ = self.__make_shared(relative_abundances, out)
-
-        row_chunks = partition_range(
-            range(len(self.species_ordering)), self.__num_processors
+            "SimilarityFromFile(similarity=%s, X=%s, chunk_size=%s",
+            similarity,
+            X,
+            chunk_size,
         )
-        args_list = [
-            (
-                chunk.start,
-                chunk.stop,
-                out_.spec,
-                self.__features.spec,
-                relative_abundances_.spec,
+        self.similarity: Callable = similarity
+        self.X: ndarray = X
+        self.chunk_size: int = chunk_size
+
+    def weighted_similarities(self, relative_abundances: ndarray) -> ndarray:
+        X_ref = put(self.X)
+        abundance_ref = put(relative_abundances)
+        futures = [
+            weighted_similarity_chunk.remote(
+                similarity=self.similarity,
+                X=X_ref,
+                relative_abundances=abundance_ref,
+                chunk_size=self.chunk_size,
+                i=i,
             )
-            for chunk in row_chunks
+            for i in range(0, self.X.shape[0], self.chunk_size)
         ]
-        with Pool(self.__num_processors) as pool:
-            pool.starmap(self.__similarity_function, args_list)
-        if out is not None and not isinstance(out, SharedArrayView):
-            out[:] = out_.data
-            return out
-        else:
-            return out_.data
+        weighted_similarity_chunks = get(futures)
+        return concatenate(weighted_similarity_chunks)
```

Comment:
One limitation of the shared memory approach that Ray uses is that the memory store is immutable so we can't write the chunks directly into a shared array. Another trade-off between memory vs simplicity (and speed, since Ray is slightly faster than the multiprocessing standard library).